src/tunnel.cc

```diff
@@ -1285,21 +1300,32 @@ TunnelStateData::noteDestinationsEnd(ErrorState *selectionError)
     // if all of them fail, tunneling as whole will fail
     Must(!selectionError); // finding at least one path means selection succeeded
 
-    if (usingDestination()) {
-        // We are already using a previously opened connection but also
-        // receiving destinations in case we need to re-forward.
-        Must(!transportWait);
+    if (transportWait) {
+        assert(!transporting());
+        notifyConnOpener();
+        return; // and continue to wait for the noteConnection() callback
+    }
+
+    if (transporting()) {
+        // We are already using a previously opened connection (but were also
+        // receiving more destinations in case we need to re-forward).
+        debugs(17, 7, "keep transporting");
         return;
     }
 
-    Must(transportWait); // or we would be stuck with nothing to do or wait for
-    notifyConnOpener();
+    // destinationsFound, but none of them worked, and we were waiting for more
+    assert(savedError);
```

Comment:
We could probably rely on the similar assertion in sendError() below.

Reply:
You are right. I am a bit worried that addressing the `XXX` below will make sendError() calls _conditional_ while this assertion (about the job state as it pertains to _this_ method) remains valid even when we cannot sendError(). I will remove this assert() if somebody insists on the removal.

Reply:
I agree with @eduard-bagdasaryan on this one.  One of the possible implementations of that XXX may may be to avoid allocating savedError objects entirely when they cannot be delivered. Whatever that ends up being this assert can be added by that change if-necessary.